rust/otap-dataflow/crates/engine/src/pipeline_metrics.rs

```diff
@@ -0,0 +1,580 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Pipeline-oriented metrics for the OTAP engine.
+//!
+//! These metrics are per pipeline instance and are sampled from the current
+//! pipeline thread. The engine uses a thread-per-core / share-nothing design:
+//! each pipeline specification instantiates one pinned thread per allocated CPU.
+//! When multiple pipeline specifications share the same CPU set, multiple pinned
+//! pipeline threads time-slice on that core (e.g. 2 pipeline specs => 2 threads
+//! pinned per allocated CPU). Each pipeline thread runs a single-threaded async
+//! runtime, so blocking operations directly affect that pipeline's metrics.
+//! The engine is NUMA-aware; these metrics help localize performance issues to a
+//! specific pipeline thread and its NUMA placement.
+//!
+//! **Metrics and interpretation**
+//! - `uptime` (`Gauge<f64>`, `{s}`):
+//!   Wall-clock time since this pipeline instance started. Useful for detecting
+//!   restarts, warm-up effects, and aligning deltas to a known lifetime.
+//!
+//! - `cpu_time` (`Counter<f64>`, `{s}`):
+//!   Cumulative CPU seconds consumed by the pipeline thread since start.
+//!   It advances only when the thread is scheduled on a CPU.
+//!   The per-interval increase (`cpu_time_delta`) is the CPU budget actually used.
+//!
+//! - `cpu_utilization` (`Gauge<f64>`, `{1}`):
+//!   Ratio of `cpu_time_delta / wall_time_delta` over the last interval,
+//!   clamped to `[0, 1]`. Interpreted as "how CPU-bound this pipeline thread is".
+//!   On a core shared by multiple pipeline specs, this reflects the pipeline's
+//!   share of that core; summing `cpu_utilization` across co-pinned pipelines
+//!   approximates total core load.
+//!
+//! - `memory_usage` (`Gauge<u64>`, `{By}`; jemalloc only):
+//!   Current heap bytes in use by the pipeline thread:
+//!   `memory_allocated - memory_freed`. A rising long-term trend indicates a leak
+//!   or persistent buffering growth inside this pipeline.
+//!
+//! - `memory_allocated` / `memory_freed` (`Gauge<u64>`, `{By}`; jemalloc only):
+//!   Cumulative heap bytes allocated/freed by this pipeline since start.
+//!   These are baselines for long-term trends and sanity checks.
+//!
+//! - `memory_allocated_delta` / `memory_freed_delta` (`Counter<u64>`, `{By}`; jemalloc only):
+//!   Per-interval heap bytes allocated/freed. They capture allocation churn
+//!   even when `memory_usage` is stable. High churn increases allocator and cache
+//!   pressure and often correlates with latency variance.
```

Comment:
:+1:

Technically these two can be used to which is equal to `memory_allocated` (i.e., Counter - Counter = UpDownCounter). We'd expect a metrics query engine to do this sort of thing!